src/ast/passes/semantic_analyser.cpp

```diff
@@ -2258,9 +2259,11 @@ void SemanticAnalyser::binop_ptr(Binop &binop)
         const auto *le = lht.GetPointeeTy();
         const auto *re = rht.GetPointeeTy();
         if (*le != *re) {
-          binop.left->addWarning(*binop.right)
-              << "comparison of distinct pointer types ('" << *le << ", '"
-              << *re << "')";
+          auto &warn = binop.addWarning();
+          warn << "comparison of distinct pointer types ('" << *le << ", '"
+               << *re << "')";
+          warn.addContext(binop.left->loc) << "left side";
+          warn.addContext(binop.right->loc) << "right side";
```

Comment:
This just to demonstrate how it can be used?

Reply:
This is to replace the existing span, which will no longer be possible.

I just threw this message together though, and we can tweak them. We probably want to attach just one context, and annotate them with the types directly. Eg. annotate the left with “invalid addition for type XXX” and annotate the right with “other type is YYY”

WDYT?

Reply:
I think we can leave the main message where it is and just add the types to the left and right e.g.
```
stdin:1:31-32: ERROR: Type mismatch for '>': comparing 'hist_t' with 'int64'
kprobe:f { @ = hist(5); if (@ > 0) { print((1)); } }

stdin:1:28-30: ERROR: left (hist_t)
kprobe:f { @ = hist(5); if (@ > 0) { print((1)); } }

stdin:1:33-34: ERROR: right (int64)
kprobe:f { @ = hist(5); if (@ > 0) { print((1)); } }

```

Reply:
Done.